enhancements/network/on-prem-dual-stack-vips.md

```diff
@@ -0,0 +1,339 @@
+will look like. We would still need compatibility logic to handle old-style
+install-configs that set the deprecated fields, but having no overlap between
+the old and new-style configs might simplify the logic somewhat.
+
+#### Add plural fields that take a list of VIPs
+```
+platform:
+  baremetal:
+    # Deprecated
+    apiVIP: 192.168.111.5
+    # Deprecated
+    ingressVIP: 192.168.111.4
+
+    apiVIPs:
+    - 192.168.111.5
+    - f00::5
+    ingressVIPs:
+    - 192.168.111.4
+    - f00::4
+```
+This is essentially a variation on the previous option where the two separate
+fields per VIP are combined into a list field with a length of up to 2. In
+this case we'd need some logic to determine whether the VIPs in the list are
+v4 or v6, but I expect that would only be a little more complicated than the
+other options. This has the advantage of being extensible if we ever wanted
+to add more addresses, but I admit to having a hard time coming up with a use
+case where more than an ipv4 and ipv6 address would be needed.
+
+### openshift/api
+
+We will need to make changes to the
+[baremetal platformStatus](https://github.com/openshift/api/blob/354aa98a475c1fcd60b41aaee735da95d7318b42/config/v1/0000_10_config-operator_01_infrastructure.crd.yaml#L311)
+to add new fields for the new VIPs. Unlike the install-config layout, I think
+it is clearly better to add discrete new fields for this feature. That way,
+backend code in machine-config-operator and baremetal-runtimecfg will not
+need to detect whether the values in the old fields are from a new (where
+the field will have to be v4) or old (where the existing field can be either
+v4 or v6) configuration.
+
+Ideally we would transition existing values to the new fields on upgrade.
+I'm unsure whether a mechanism exists for us to do that, however. If not, then
+some logic (probably in MCO) will be needed to map values from the old fields
+to the appropriate ip version.
+
+The new structure of platformStatus would look like this:
+
+```
+baremetal:
+  description: BareMetal contains settings specific to the BareMetal platform.
+  type: object
+  properties:
+    # Deprecated
+    apiServerInternalIP:
+      description: apiServerInternalIP is an IP address...
+      type: string
+    apiServerInternalIPv4:
+      description: apiServerInternalIPv4 is an IP address...
+      type: string
+    apiServerInternalIPv6:
+      description: apiServerInternalIPv6 is an IP address...
+      type: string
+    # Deprecated
+    ingressIP:
+      description: ingressIP is an external IP...
+      type: string
+    ingressIPv4:
+      description: ingressIPv4 is an external IP...
+      type: string
+    ingressIPv6:
+      description: ingressIPv6 is an external IP...
+      type: string
+    nodeDNSIP:
+    ...
+```
+
+### machine-config-operator
+
+We will need to wire in the new VIPs to the manifests and configuration
+templates. There should be conditionals around the v4 and v6 sections
```

Comment:
So if you're going to be fiddling with all this stuff anyway [we may be able to fix UPI node IP detection too](https://bugzilla.redhat.com/show_bug.cgi?id=1949827#c59):

> Currently, for UPI, if the MCO knows the apiserver IP at install time, it will write it into the nodeip-configuration service, and so we will pick the IP with the most direct route to the apiserver (just like in the IPI case). However, in some cases (notably vSphere UPI), the MCO does not know the apiserver IP at install time and thus cannot do this. From what I can tell though, the apiserver IP is still *known* (eg, to the installer) in this case, it's just not recorded anywhere that the MCO can see at install time. So if we fix the install-time config plumbing so that the MCO always has access to the configured apiserver IP, then it should always be able to pass that to nodeip-configuration, and we should always get the right IP. (So this probably requires one or more additions to openshift/api, plus changes to openshift/installer to fill in the new API, and changes to openshift/machine-config-operator to consume the new API. It should not require any changes to openshift/baremetal-runtimecfg.)

I haven't dug too deeply into this but it seems like while we may need to continue having platform-specific `spec` VIP fields, we might be able to have platform-_independent_ `status` VIP fields? (Which would really simplify [this](https://github.com/openshift/machine-config-operator/blob/bbe78cfe1eb9bfa9dc5e8fe27fb8155e648dd848/pkg/operator/render.go#L258).)

Reply:
Hmm, that may be possible. I'm not familiar with UPI installer data, but if it's just adding a field somewhere it's probably doable.